WcaOnRails/app/javascript/edit-schedule/EditSchedule.jsx

```diff
@@ -0,0 +1,218 @@
+import React from 'react'
+import cn from 'classnames'
+import _ from 'lodash'
+import {
+  Alert,
+  Col,
+  Panel,
+  PanelGroup,
+  Row,
+} from 'react-bootstrap'
+
+import { rootRender, promiseSaveWcif } from 'edit-schedule'
+import { EditVenue } from './EditVenue'
+import { SchedulesEditor } from './SchedulesEditor'
+import { initElementsIds, newVenueId } from './utils'
+
+export const schedulesEditPanelSelector = "#schedules-edit-panel";
+
+export default class EditSchedule extends React.Component {
+  componentWillMount() {
+    this.setState({ savedScheduleWcif: _.cloneDeep(this.props.competitionInfo.scheduleWcif) });
```

Comment:
In practice, updating pieces of a deep javascript object can be a bit of a pain without mutation, and this `cloneDeep` is an easy way to avoid mutating the original object. If someone is motivated to look into something like Immutable.js some day, maybe that would be good alternative =)

Reply:
Actually, it's not that painful ^^ When I found it too cumbersome, I started with [a helper function](https://github.com/jonatanklosko/groupifier/blob/b11e6a5c75f5ffd5d13af6241100dedbacd830a3/src/logic/helpers.js#L1-L12). Then you can use it to define more specific ones as needed (like `setIn` and `mergeIn` below). This way you get a modified copy, but it shares many inner objects with the original object. Anyway, I don't have much React experience yet, just pointing it out =)